### PR TITLE
Add hmac_sha1 Presto Functions

### PR DESCRIPTION
**Pull request by frankliee**
Add [hmac_sha1](https://prestodb.io/docs/current/functions/binary.html#:~:text=given%20key.-,hmac_sha1,-(binary%2C)) Presto Functions

```
hmac_sha1(binary, key) → varbinary
     Computes HMAC with sha1 of binary with the given key.
```

### CODE REVIEW

File: velox/functions/prestosql/StringFunctions.h [hunk 1]
```diff
@@ -120,6 +120,18 @@ struct Sha512Function {
...
 
+/// hmac_sha1(varbinary) -> varbinary
+template <typename T>
+struct HmacSha1Function {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  template <typename TOuput, typename TInput>
+  FOLLY_ALWAYS_INLINE void
+  call(TOuput& result, const TInput& data, const TInput& key) {
+    return stringImpl::HmacSha1(result, key, data);
```
Comment (xiaoxmeng) [inline velox/functions/prestosql/StringFunctions.h:131]: call return void?
```
s/return stringImpl::HmacSha1(result, key, data);/stringImpl::HmacSha1(result, key, data);/
```

File: velox/functions/prestosql/tests/StringFunctionsTest.cpp [hunk 2]
```diff
@@ -1201,6 +1201,34 @@ TEST_F(StringFunctionsTest, sha512) {
...
+  EXPECT_EQ(
+      hexToDec("d49c944625bdde6c47ad93ea63952bfcf16a630a"),
+      hmacSha1("hashme", "velox"));
+  EXPECT_EQ(
+      hexToDec("c19b6b753fe4ac28579c7e84d18feb29760a0d1c"),
+      hmacSha1("Infinity", "velox"));
+  EXPECT_EQ(
+      hexToDec("d0569c4a4f3df995b04ec497b12872c4a2f97517"),
+      hmacSha1("", "velox"));
+  EXPECT_EQ(std::nullopt, hmacSha1(std::nullopt, "velox"));
```
Comment (xiaoxmeng) [inline velox/functions/prestosql/tests/StringFunctionsTest.cpp:1229]: Can key be null or empty string?

File: velox/functions/prestosql/StringFunctions.h (hunk 1 above)
Comment (frankliee) [reply velox/functions/prestosql/StringFunctions.h:131]: I see https://github.com/facebookincubator/velox/pull/3406 has changed the return type of HmacSha256Function to `void`, so I Modified it in advance,

File: velox/functions/prestosql/tests/StringFunctionsTest.cpp (hunk 2 above)
Comment (frankliee) [reply velox/functions/prestosql/tests/StringFunctionsTest.cpp:1229]: It just to check the cover case.
I see the other tests in StringFunctionsTest.cpp have contain null and "".

Comment (xiaoxmeng) [reply velox/functions/prestosql/tests/StringFunctionsTest.cpp:1229]: Can we add to test value is empty or null case as well? Thanks!

**Review by xiaoxmeng — APPROVED** [review]

File: velox/functions/lib/string/StringImpl.h [hunk 3]
```diff
@@ -288,6 +288,17 @@ FOLLY_ALWAYS_INLINE void sha512(TOutString& output, const TInString& input) {
       folly::ByteRange((const uint8_t*)input.data(), input.size()));
 }
 
+// Compute the HMAC-SHA1 Hash.
+template <typename TOutString, typename TInString>
+FOLLY_ALWAYS_INLINE void
+HmacSha1(TOutString& output, const TInString& key, const TInString& data) {
```
Comment (aditi-pandit) [inline velox/functions/lib/string/StringImpl.h:294]: Nit : Please rename this function and the one below as hmacSha1 and hmacSha256 to be consistent with the function naming.

File: velox/functions/prestosql/tests/StringFunctionsTest.cpp [hunk 4]
```diff
@@ -1201,6 +1201,34 @@ TEST_F(StringFunctionsTest, sha512) {
...
+  //         print(hmac.new(key, data, digestmod='sha1').hexdigest())
+  // >>> sha1(b"hashme", b"velox")
+  // d49c944625bdde6c47ad93ea63952bfcf16a630a
+  // >>> sha1(b"Infinity", b"velox")
+  // c19b6b753fe4ac28579c7e84d18feb29760a0d1c
+  // >>> sha1(b"", b"velox")
+  // d0569c4a4f3df995b04ec497b12872c4a2f97517
+  EXPECT_EQ(
+      hexToDec("d49c944625bdde6c47ad93ea63952bfcf16a630a"),
+      hmacSha1("hashme", "velox"));
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/StringFunctionsTest.cpp:1222]: Please can we also add tests for large input strings ( > 12 chars to test out of line data)

**Comment by aditi-pandit** [comment]
Thanks for adding this function.

**Review by xiaoxmeng — COMMENTED** [review]
@frankliee there is a conflict and please rebase. Thank you!

**Comment by frankliee** [comment]
> @frankliee there is a conflict and please rebase. Thank you!

updated

File: velox/functions/prestosql/tests/StringFunctionsTest.cpp (hunk 4 above)
Comment (frankliee) [reply velox/functions/prestosql/tests/StringFunctionsTest.cpp:1224]: added   hmacSha1("12345abcde54321", "velox"));

File: velox/functions/lib/string/StringImpl.h (hunk 3 above)
Comment (frankliee) [reply velox/functions/lib/string/StringImpl.h]: done

File: velox/functions/prestosql/tests/StringFunctionsTest.cpp (hunk 2 above)
Comment (frankliee) [reply velox/functions/prestosql/tests/StringFunctionsTest.cpp:1231]: I still don't get it.
There already have "" and null.

```
  EXPECT_EQ(
      hexToDec("d0569c4a4f3df995b04ec497b12872c4a2f97517"),
      hmacSha1("", "velox"));
  EXPECT_EQ(std::nullopt, hmacSha1(std::nullopt, "velox"));
```

Comment (xiaoxmeng) [reply velox/functions/prestosql/tests/StringFunctionsTest.cpp:1231]: Sorry I mean
EXPECT_EQ( hexToDec("d0569c4a4f3df995b04ec497b12872c4a2f97517"), hmacSha1("velox", ""));
EXPECT_EQ(std::nullopt, hmacSha1("velox", std::nullopt));

**Review by aditi-pandit — APPROVED** [review]

File: velox/functions/prestosql/tests/StringFunctionsTest.cpp (hunk 2 above)
Comment (frankliee) [reply velox/functions/prestosql/tests/StringFunctionsTest.cpp:1233]: Done and fix a test fail due to illegal character

Comment (xiaoxmeng) [reply velox/functions/prestosql/tests/StringFunctionsTest.cpp:1233]: Thanks!
